Enforce assert equivalence in Z3 verifier before claiming passes are safe

DeeveshChowdary opened the pull request:
While looking into https://github.com/google/xls/issues/3293, I realized the Z3 equivalence checker never even sees assert behavior.. it either ignores them or strips them if you pass `--ignore_asserts`. That meant passes could tighten/loosen asserts, or add new ones, and the proof would still come back green because we never asked whether the asserts themselves still fired on the same inputs... I fixed this by making `TryProveEquivalence` pre-process both versions: it now pairs asserts by label/original_label, proves each pair’s condition matches, and forces any unmatched assert to be a tautology. Those checks are ANDed into the predicate we hand to Z3, so `ScopedVerify*` and `check_ir_equivalence_main` pick up the safety net automatically.

Fixes: https://github.com/google/xls/issues/3293

allight commented on xls/solvers/z3_ir_equivalence.cc:
```diff
@@ -46,6 +49,92 @@ namespace xls::solvers::z3 {
+          original_assert->operand(Assert::kConditionOperand);
+      Node* cond_b =
+          transformed_it->second->operand(Assert::kConditionOperand);
+      checks.push_back(f->AddNode(std::make_unique<CompareOp>(
```
We usually add new nodes like

```
XLS_ASSIGN_OR_RETURN(std::back_inserter(checks), f->MakeNode<CompareOp>(SourceInfo, cond_a, cond_b, Op::Eq));
```

or

```
XLS_ASSIGN_OR_RETURN(std::back_inserter(checks), f->MakeNodeWithName<CompareOp>(SourceInfo, cond_a, cond_b, Op::Eq, absl::StrCat("assert_", label, "_match")));
```

here and below.

The MakeNode path will also invoke the verifier to do some basic structural checks of the node being added which can catch some bugs.

allight commented on xls/solvers/z3_ir_equivalence.cc:
```diff
@@ -46,6 +49,92 @@ namespace xls::solvers::z3 {
+  if (checks.size() == 1) {
+    return checks.front();
+  }
+  return f->AddNode(std::make_unique<NaryOp>(
```
As above use MakeNode

DeeveshChowdary commented on xls/solvers/z3_ir_equivalence.cc:
```diff
@@ -46,6 +49,92 @@ namespace xls::solvers::z3 {
+    auto transformed_it = transformed_asserts.find(label);
+    if (transformed_it != transformed_asserts.end()) {
+      Node* cond_a =
+          original_assert->operand(Assert::kConditionOperand);
```
Done. BuildAssertChecks now uses Assert::condition() for both the original and transformed nodes.

DeeveshChowdary commented on xls/solvers/z3_ir_equivalence.cc:
```diff
@@ -46,6 +49,92 @@ namespace xls::solvers::z3 {
+}
+
+Node* CombineChecks(Function* f, absl::Span<Node* const> checks) {
+  XLS_DCHECK(!checks.empty());
```
Updated. CombineChecks now returns StatusOr<Node*> and uses XLS_RET_CHECK(!checks.empty()).

DeeveshChowdary commented on xls/solvers/z3_ir_equivalence_test.cc:
```diff
@@ -427,5 +427,72 @@ TEST_F(EquivalenceTest, MultiFunctionIgnoresAssert) {
               IsOkAndHolds(IsProvenTrue()));
 }
 
+TEST_F(EquivalenceTest, DetectsMismatchedAssertCondition) {
```
Done.

DeeveshChowdary commented:
Thanks @ericastor and @allight! I updated the code. Please take another look whenever possible.

> Also the top comment of the PR thread becomes the merge commits message. If you add `Fixes: https://github.com/google/xls/issues/3293` to the end of it then that bug will be automatically closed when merged.

I see.. I've added it to the comment now. Thanks again!
